### PR TITLE
Refactored waja to add referral during visit

### PR DESCRIPTION
**Pull request by nKataraia**

### CODE REVIEW

File: opensrp-chw/build.gradle [hunk 1]
```diff
@@ -249,14 +250,15 @@ dependencies {
...
-    implementation('org.smartregister:opensrp-client-chw-referral:1.3.13-NACP-ALPHA-SNAPSHOT@aar') {
+    implementation('org.smartregister:opensrp-client-chw-referral:1.3.13-NACP-LOCAL-ALPHA-SNAPSHOT@aar') {
         transitive = true
         exclude group: 'org.smartregister', module: 'opensrp-client-core'
         exclude group: 'org.smartregister', module: 'opensrp-client-native-form'
         exclude group: 'org.smartregister', module: 'opensrp-client-configurable-views'
         exclude group: 'com.android.support', module: 'appcompat-v7'
         exclude group: 'id.zelory', module: 'compressor'
         exclude group: 'com.rengwuxian.materialedittext', module: 'library'
+        exclude group: 'com.ibm.fhir', module: 'fhir-path'
```
Comment (gosso22) [inline opensrp-chw/build.gradle]: @nKataraia - I think this has been fixed so we dont need to exclude fhir-path library and include the version below.

File: opensrp-chw/build.gradle [hunk 2]
```diff
@@ -249,14 +250,15 @@ dependencies {
 //        exclude group: 'id.zelory', module: 'compressor'
 //    }
 
-    implementation('org.smartregister:opensrp-client-chw-referral:1.3.13-NACP-ALPHA-SNAPSHOT@aar') {
+    implementation('org.smartregister:opensrp-client-chw-referral:1.3.13-NACP-LOCAL-ALPHA-SNAPSHOT@aar') {
```
Comment (gosso22) [inline opensrp-chw/build.gradle]: @nKataraia - did you make any changes to the referral library? If not probably just leave this to the version others are using locally.

File: opensrp-chw/src/nacp/java/org/smartregister/chw/interactor/ChildHomeVisitInteractorFlv.java [hunk 3]
```diff
@@ -16,42 +18,53 @@
...
-            evaluateVitaminA(serviceWrapperMap);
-            evaluateDeworming(serviceWrapperMap);
-            evaluateMalariaPrevention();
-            evaluateCounselling();
-            evaluateNutritionStatus();
-            evaluateObsAndIllness();
-            evaluateMalnutritionScreening(serviceWrapperMap);
-            evaluateProblemSolving();
+            this.serviceWrapperMap=serviceWrapperMap;
+            evaluateToddlerDanger();
```
Comment (gosso22) [inline opensrp-chw/src/nacp/java/org/smartregister/chw/interactor/ChildHomeVisitInteractorFlv.java]: @nKataraia - evaluating other action inside danger signs makes danger signs mandatory, for the cases where there is no danger signs screening then no modules show with this implementation. Can you check out this scenario?

File: opensrp-chw/src/test/java/org/smartregister/chw/Futa.java
Comment (gosso22) [inline opensrp-chw/src/test/java/org/smartregister/chw/Futa.java:1]: @nKataraia - maybe you should remove this

File: opensrp-chw/src/nacp/java/org/smartregister/chw/util/BoolEvaluator.java
Comment (gosso22) [inline opensrp-chw/src/nacp/java/org/smartregister/chw/util/BoolEvaluator.java:1]: @nKataraia - are you using this?

**Review by gosso22 — CHANGES_REQUESTED** [review]
@nKataraia - I have left you a couple of comments for this PR, overall this is an awesome implementation. It is fun to review :). I think you might need to look into Child and PNC. For child the danger signs is made to be a condition for other modules, so for the case of children who dont have danger signs screening (for example children older than 12 months) the visit doesnt show any module. For PNC, if you select danger signs, then go back and select none. it still shows refer facility module. I think this should disappear and show other modules.

Thanks for working on this.

**Review by gosso22 — APPROVED** [review]

File: opensrp-chw/src/nacp/java/org/smartregister/chw/interactor/AncHomeVisitInteractorFlv.java [hunk 4]
```diff
@@ -100,6 +104,31 @@ private void evaluateDangerSigns(Map<String, List<VisitDetail>> details,
...
+    private void evaluateFacilityReferral(String referralPayload,Map<String, List<VisitDetail>> details,
+                                          final Context context) throws BaseAncHomeVisitAction.ValidationException {
+        String formName="referral_facility_selection";
+        JSONObject jsonForm = FormUtils.getFormUtils().getFormJson(formName);
+        Map<String,String> facilityOptions= LocationUtils.INSTANCE.getFacilitiesKeyAndName();
+        JsonFormUtilsFlv.overwriteQuestionOptions("chw_referral_hf",facilityOptions, jsonForm);
+
+        if (details != null) ChwAncJsonFormUtils.populateForm(jsonForm, details);
+
+        JSONObject referralProblem=FacilitySelectionActionHelper.copyReferralProblem(referralPayload,"danger_signs_present");
```
Comment (issyzac) [inline opensrp-chw/src/nacp/java/org/smartregister/chw/interactor/AncHomeVisitInteractorFlv.java:116]: Why do we have this implementation inside the action helper?

**Review by issyzac — APPROVED** [review]

**Pull request merged**